flink-runtime/src/main/java/org/apache/flink/runtime/io/AsyncDataInput.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Interface defining couple of essential methods for asynchronous and non blocking data polling.
```

Comment:
I would suggest to outline the intended usage here in the class level doc:
Consume via poll until you get some representation of null or empty, check if the data is exhausted and if not wait on the future. This might be helpful, especially because otherwise this looks pretty similar to an iterator, but can be used in a slightly different and more efficient way w.r.t. `isFinished()/hasNext()`